### DISCUSSION TITLE
Unable To Get LEGO Spike Prime To Recognize pybricks commands via Bluetooth

### DISCUSSION THREAD

**Discussion by jfantell** [pybricks/support#1352] [Q&A]
Hello! :)

I'm new to Pybricks and LEGO Spike Prime.

I'm trying to control a LEGO Spike Prime via a Python 3 Script running on my Mac.

The code snippet I've been using is provided below.

Note: Before the script starts running, the blue button light on the hub is blinking.

Expected Result: The light matrix on the top of the hub blinks; the motor starts running, and then it stops 10 seconds later.
Actual Result: The bluetooth connection is presumably established -- per a log message in the code -- and the blue button light on the hub stops blinking (stays solid blue).

```
import asyncio
from bleak import BleakClient
from pybricks.pupdevices import Motor
from pybricks.parameters import Port
from pybricks.hubs import PrimeHub
import time

# Replace with your PrimeHub's Bluetooth address
HUB_ADDRESS = "5ADF50CA-C49F-B56E-1B15-AFB84A01B177"

# Initialize the PrimeHub and motor variables globally
hub = None
motor = None

async def run_motor():
    client = BleakClient(HUB_ADDRESS)
    try:
        # Connect to the hub
        await client.connect()
        if client.is_connected():
            print("Connected to the hub!")

            # Initialize the hub and motor within the connected context
            global hub, motor
            hub = PrimeHub(client)

            # Hub Name
            print(hub.system.name())

            # Light matrix
            hub.light.blink(11, [10, 100])

            motor = Motor(Port.A, client)  # Replace 'A' if your motor is on a different port

            print("Attempting to start motor...")
            motor.run(-500)
            print("Motor should be running.")

            # Wait for 10 seconds
            await asyncio.sleep(10)

            # Stop the motor
            motor.stop()
            print("Motor operation complete.")

        else:
            print("Failed to connect to the hub.")
    except Exception as e:
        print(f"An error occurred: {e}")
    finally:
        await client.disconnect()

# Run the async function
asyncio.run(run_motor())
```
I've also tried using the code tutorial found here to no avail: https://pybricks.com/projects/tutorials/wireless/hub-to-device/.

I just want to get the Hub/Motor recognizing commands via a bluetooth connection. Please advise. Thanks in advanced!!

Additional Information:

* Python 3.9.6 script running on MacOS 14.0
* LEGO Spike Prime (Model HUB NO. 6)
* Pybricks Python Package v3.3.0
* Pybricks Firmware (installed on the Hub via https://code.pybricks.com/): Version ? (whichever version is currently on the website)

**Comment by dlech** [comment]
The `pybricks` package on PyPI is just stubs for code completion and doesn't actually do anything when run on a PC. You need to write a separate program that runs on the hub, then download and run that program. Then you can use Bleak or pybricksdev to communicate with the hub running that program.

**Comment by jfantell** [comment]
Thanks for the guidance but I'm still a bit confused. I'm trying to write a program that will start and stop the LEGO Spike Prime remotely based on certain conditions. 

So when you say "You need to write a separate program that runs on the hub, then download and run that program"  are you suggesting writing a program that runs on the hub that basically acts as a "server" with endpoints that trigger the motor (connected to the hub) to to start and stop? And then invoke these endpoints from a script running on a PC? Are there any particular examples or tutorials on this topic for me to reference?

**Reply by dlech** [reply]
> are you suggesting writing a program that runs on the hub that basically acts as a "server" with endpoints that trigger the motor (connected to the hub) to to start and stop? And then invoke these endpoints from a script running on a PC?

Yes

> Are there any particular examples or tutorials on this topic for me to reference?

https://pybricks.com/projects/tutorials/ (although these don't use `pybricksdev`)

https://github.com/Novakasa/brickrail/tree/master/ble-server (a very complex example of  a project that uses pybricksdev)

**Reply by Novakasa** [reply]
I also wrote this example once, which is a minimal example of one-directional communication from a PC script to a pybricks hub: https://github.com/pybricks/support/issues/470#issuecomment-917070340
A few comments below that is a more involved example with bidirectional communication. There have been some pybricksdev API changes since but overall it should still work something like that.

**Reply by Novakasa** [reply]
Otherwise, in the project @dlech linked, the files to look at are `ble-server/ble_hub.py` and `ble-server/hub_programs/io_hub_unfrozen.py`. This however is quite involved with a custom communication protocol with error checking, for simple remote control this is probably overkill.

There is a test script that shows off the API in [`ble-server/test_io_hub.py`](https://github.com/Novakasa/brickrail/blob/master/ble-server/test_io_hub.py)
